File: javascript/exercise40/filtering-records.js
import Database from "better-sqlite3";
import { input } from "pynput";

const db = new Database("./records.db", { fileMustExist: true });

const search = await input("Enter a search string: ");

/** @type [{
 * first_name: string,
 * last_name: string,
 * position: string,
 * date: string
 * }] rows */
const rows = db
	.prepare("SELECT * FROM employees WHERE first_name like ?")
	.all(`${search}%`);
db.close();

// console.log(rows)

console.log(
	`${"Name".padEnd(20)} | ${"Position".padEnd(20)} | Separation Date`,
);
console.log("-".repeat(60));
for (const row of rows) {
	const fullname = `${row.first_name} ${row.last_name}`.padEnd(20);
	const position = row.position.padEnd(20);
	const date = row.date || "";

	console.log(`${fullname} | ${position} | ${date}`);
}

File: python/filtering_records/filtering_records.py
import sqlite3
from typing import Any
# from typing import Any


with sqlite3.connect("./employees.db") as connection:
    cursor = connection.cursor()

search = input("Enter a search string: ")

rows = cursor.execute(
    "SELECT * FROM employees WHERE first_name LIKE ? or last_name LIKE ?",
    (f"%{search}%", f"%{search}%"),
).fetchall()


employees: list[dict[str, Any]] = []

for name, surname, position, date in rows:
    employees.append(
        {
            "name": name,
            "surname": surname,
            "position": position,
            "date": date,
        }
    )

print("Results:")
print(f"{'Name':20}| {'Position':12}| Separation Date")
print(f"{'-'*20}| {'-'*12}| {'-'*15}")
for emp in employees:
    name = f"{emp["name"]} {emp["surname"]}"
    print(f"{name:20}| {emp["surname"]:12}| {emp["date"] or ''}")
